File: examples/chapter11/eulerheatstab.m
%%
% Both time stepping methods solved
% $\mathbf{u}'=\mathbf{D}_{xx}\mathbf{u}$, for the matrix
m = 40;  [~,~,Dxx] = diffper(m,[0,1]);

%%
% The eigenvalues of this martrix are real and negative.
lambda = eig(Dxx);
plot(real(lambda),imag(lambda),'.')
axis equal, grid on   % ignore this line
xlabel('Re \lambda'),  ylabel('Im \lambda')   % ignore this line
title('Eigenvalues')   % ignore this line

%%
% The Euler method is absolutely stable in the region $|\zeta+1|<1$ in the
% complex plane:
phi = linspace(0,2*pi,361);
z = exp(1i*phi) - 1;   % unit circle shifted to the left by 1
fill(real(z),imag(z),[.8 .8 1])
axis equal, grid on   % ignore this line
xlabel('Re \lambda'),  ylabel('Im \lambda')   % ignore this line
title('Stability region')   % ignore this line

%%
% In order to get inside this region, we have to find $\tau$ such
% that $\lambda \tau > -2$ for all eigenvalues $\lambda$. This is an \emph{upper} bound on
% $\tau$. 
lambda_min = min(lambda)
max_tau = -2 / lambda_min

%%
% Here we plot the resulting values of $\zeta=\lambda \tau$. 
zeta = lambda*max_tau;
hold on
plot(real(zeta),imag(zeta),'.')
title('Stability region and \zeta values')   % ignore this line

%%
% In backward Euler, the region is $|\zeta-1|>1$. Because they are all on
% the negative real axis, all of the $\zeta$ values will fit no matter what
% $\tau$ is chosen.
clf,  fill([-6 6 6 -6],[-6 -6 6 6],[.8 .8 1]),  hold on
z = exp(1i*phi) + 1;   % unit circle shifted right by 1
fill(real(z),imag(z),'w')
plot(real(zeta),imag(zeta),'.')
axis equal, axis([-4 2 -3 3]), grid on   % ignore this line
xlabel('Re \lambda'),  ylabel('Im \lambda')   % ignore this line
title('Stability region and \zeta values')   % ignore this line

File: julia/scripts/chapter11/eulerheatstab.jl

include("../FNC.jl")
m = 40;  Dxx = FNC.diffper(m,[0,1])[3];

using LinearAlgebra
lambda = eigvals(Dxx)

using Plots
scatter(real(lambda),imag(lambda),m=2,
    xaxis=("Re \\lambda"),  yaxis=("Im \\lambda"),
    title="Eigenvalues",leg=:none)

phi = 2pi*(0:360)/360
z = @.exp(1im*phi) - 1;   # unit circle shifted to the left by 1

plot(Shape(real(z),imag(z)),color=RGB(.8,.8,1),aspect_ratio=1,
    xaxis=("Re \\lambda"),  yaxis=("Im \\lambda"),
    title="Stability region",leg=:none) 

lambda_min = minimum(lambda)
@show max_tau = -2 / lambda_min;

zeta = lambda*max_tau
scatter!(real(zeta),imag(zeta),
    title="Stability region and \\zeta  values")

plot(Shape([-6,6,6,-6],[-6,-6,6,6]),color=RGB(.8,.8,1),aspect_ratio=1)
z = @.exp(1im*phi) + 1;   # unit circle shifted right by 1
plot!(Shape(real(z),imag(z)),color=:white)
scatter!(real(zeta),imag(zeta),
    xaxis=([-4,2],"Re \\lambda"),  yaxis=([-3,3],"Im \\lambda"),
    title="Stability region and \\zeta values",leg=:none)
